runtime/gc_glue_java/MetronomeDelegate.cpp

```diff
@@ -1115,5 +1133,396 @@ MM_MetronomeDelegate::releaseExclusiveVMAccess(MM_EnvironmentBase *env, bool rel
+	 * them is scanned and then allowed to execute, any object it takes off the finalizer
+	 * queue had better also be scanned.  An alternative would be to put a special read
+	 * barrier in the queue-removal action but controlling the order is an easy solution.
+	 *
+	 * It is also important to scan JNI global references after scanning threads, because
+	 * the JNI global reference barrier is executed at deletion time, not creation time.
+	 * We could have barriers in both, but controlling the order is an easy solution.
+	 *
+	 *
+	 * The Metronome write barrier ensures that no unscanned thread can expose an object
+	 * to other threads without it becoming a root and no scanned thread can make an
+	 * object that was once reachable unreachable until it has been traced ("snapshot at
+	 * the beginning with a fuzzy snapshot").  This eliminates other order dependencies
+	 * between portions of the scan or requirements that multiple phases be done as an
+	 * atomic unit.  However, some phases are still done atomically because we have not
+	 * yet determined whether the iterators that they use are safe and complete and have
+	 * not even analyzed in all cases whether correctness depends on completeness.
+	 */
+	if (env->_currentTask->synchronizeGCThreadsAndReleaseMaster(env, UNIQUE_ID)) {
+#if defined(J9VM_GC_FINALIZATION)
+		/* Note: if iterators are safe in scanFinalizableObjects, disableYield() could be
+		 * removed.
+		 */
+		env->disableYield();
+		rootMarker.scanFinalizableObjects(env);
+		env->enableYield();
+		_scheduler->condYieldFromGC(env);
+#endif /* J9VM_GC_FINALIZATION */
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		if (!isDynamicClassUnloadingEnabled()) {
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+			/* We are scanning all classes, no need to include stack frame references */
+			rootMarker.setIncludeStackFrameClassReferences(false);
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+		} else {
+			rootMarker.setIncludeStackFrameClassReferences(true);
+		}
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+		env->_currentTask->releaseSynchronizedGCThreads(env);
+	}
+
+	rootMarker.scanThreads(env);
+
+	if (env->_currentTask->synchronizeGCThreadsAndReleaseMaster(env, UNIQUE_ID)) {
+		_extensions->newThreadAllocationColor = GC_MARK;
+		_realtimeGC->disableDoubleBarrier(env);
+		if (_realtimeGC->verbose(env) >= 3) {
+			rootMarker.reportThreadCount(env);
+		}
+
+		/* Note: if iterators are safe for some or all remaining atomic root categories,
+		 * disableYield() could be removed or moved inside scanAtomicRoots.
+		 */
+		env->disableYield();
+		rootMarker.scanAtomicRoots(env);
+		env->enableYield();
+		rootMarker.scanIncrementalRoots(env);
+	
+		env->_currentTask->releaseSynchronizedGCThreads(env);
+	}
+
+	env->setRootScanner(NULL);
+}
+
+void
+MM_MetronomeDelegate::markLiveObjectsScan(MM_EnvironmentRealtime *env)
+{
+	MM_RealtimeMarkingSchemeRootClearer rootScanner(env, _realtimeGC);
```

Comment:
Does not look right. Clearable roots scanning belongs to complete phase. Again, look at the base MarkingScheme.

Reply:
Again, if you look at the base class, it has no such code in the complete phase. There are no clearable roots in any phase. I'm not sure why you think its an example I can follow. Perhaps you are thinking of another class somewhere else entirely.

void MM_MarkingScheme::markLiveObjectsComplete(MM_EnvironmentBase *env)
{
	_delegate.workerCompleteGC(env);
}

MMINLINE void MM_MarkingDelegate::workerCompleteGC(MM_EnvironmentBase *env)
{
		/* All threads flush buffers before this point, and complete any remaining language-specific marking tasks */
		if (env->_currentTask->synchronizeGCThreadsAndReleaseSingleThread(env, UNIQUE_ID)) {
			/* Perform single-threaded tasks here */
			env->_currentTask->releaseSynchronizedGCThreads(env);
		}
}

Reply:
Moving the root clearer code to markLiveObjectsRoots, as well, would change the original order of the code. I'm worried that will break it

Reply:
It does have it - not sure why you did not see it: https://github.com/eclipse/openj9/blob/master/runtime/gc_glue_java/MarkingDelegate.cpp#L158